autoprecompiles/src/range_constraint_optimizer.rs

```diff
@@ -0,0 +1,146 @@
+    /// constraint if both values are unknown (because the allowed values of `a`
+    /// depend on `b`), but it is a pure range constraint if `b` is known.
+    /// Any stateful bus interaction is not a pure range constraint.
+    fn pure_range_constraints<V: Ord + Clone + Eq + Display + Hash>(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Option<RangeConstraintMap<T, V>>;
+
+    /// Given a set of range constraints, returns a list of bus interactions
+    /// that implements them. The implementation is free to implement multiple
+    /// range constraints using a single bus interaction.
+    /// As all input range constraints are unconditional, the multiplicity of
+    /// the returned bus interactions should be 1.
+    fn batch_make_range_constraints<V: Ord + Clone + Eq + Display + Hash>(
+        &self,
+        range_constraints: RangeConstraintMap<T, V>,
+    ) -> Vec<BusInteraction<GroupedExpression<T, V>>>;
+}
+
+/// Optimizes range constraints, minimizing the number of bus interactions. This step removes range
+/// constraints that are already implied by existing constraints, and it implements bit constraints
+/// via polynomial constraints, if the degree bound allows.
+pub fn optimize_range_constraints<T: FieldElement, V: Ord + Clone + Hash + Eq + Display>(
+    mut system: ConstraintSystem<T, V>,
+    bus_interaction_handler: impl BusInteractionHandler<T> + RangeConstraintHandler<T> + Clone,
+    degree_bound: DegreeBound,
+) -> ConstraintSystem<T, V> {
+    // Remove all pure range constraints, but collect what was removed.
+    let mut to_constrain = BTreeMap::new();
+    system.bus_interactions.retain(|bus_int| {
+        if bus_int.multiplicity != GroupedExpression::from_number(T::one()) {
+            // Most range constraints are unconditional in practice, it's probably not
+            // worth dealing with the conditional ones.
+            return true;
+        }
+
+        match bus_interaction_handler.pure_range_constraints(bus_int) {
+            Some(range_constraints) => {
+                for (expr, rc) in range_constraints {
+                    let existing_rc = to_constrain
+                        .entry(expr)
+                        .or_insert_with(RangeConstraint::default);
+                    *existing_rc = existing_rc.conjunction(&rc);
+                }
+                false
+            }
+            None => true,
+        }
+    });
+
+    // Filter range constraints that are already implied by existing constraints.
+    // TODO: They could also be implied by each other.
+    let mut solver = new_solver(system.clone(), bus_interaction_handler.clone());
+    solver.solve().unwrap();
+    let to_constrain = to_constrain
+        .into_iter()
+        .filter(|(expr, rc)| {
+            let current_rc = solver.range_constraint_for_expression(expr);
+            current_rc != current_rc.conjunction(rc)
```

Comment:
So this keeps only range constraints that are not already implied but the polynomial constrains + remaining bus interactions that were removed above?